sqlserver/datadog_checks/sqlserver/schemas.py

```diff
@@ -0,0 +1,365 @@
+            "collection_interval": schemas_collection_interval,
+            "dbms_version": None,
+            "tags": self._check.non_internal_tags,
+            "cloud_metadata": self._check._config.cloud_metadata,
+        }
+        self._dataSubmitter = SubmitData(self._check.database_monitoring_metadata, base_event, self._log)
+
+    """Collects database information and schemas and submits to the agent's queue as dictionaries
+    schema dict
+    key/value:
+        "name": str
+        "id": str
+        "owner_name": str
+        "tables" : list of tables dicts
+            table
+            key/value:
+                "id" : str
+                "name" : str
+                columns: list of columns dicts
+                    columns
+                    key/value:
+                        "name": str
+                        "data_type": str
+                        "default": str
+                        "nullable": bool
+            indexes : list of index dicts
+                index
+                key/value:
+                    "name": str
+                    "type": str
+                    "is_unique": bool
+                    "is_primary_key": bool
+                    "is_unique_constraint": bool
+                    "is_disabled": bool,
+                    "column_names": str
+            foreign_keys : list of foreign key dicts
+                foreign_key
+                key/value:
+                    "foreign_key_name": str
+                    "referencing_table": str
+                    "referencing_column": str
+                    "referenced_table": str
+                    "referenced_column": str
+            partitions: partition dict
+                partition
+                key/value:
+                    "partition_count": int
+    """
+
+    @tracked_method(agent_check_getter=agent_check_getter)
+    def collect_schemas_data(self):
+        self._dataSubmitter.reset()
+        self._dataSubmitter.set_base_event_data(
+            self._check.resolved_hostname,
+            self._check.non_internal_tags,
+            self._check._config.cloud_metadata,
+            "{},{}".format(
+                self._check.static_info_cache.get(STATIC_INFO_VERSION, ""),
+                self._check.static_info_cache.get(STATIC_INFO_ENGINE_EDITION, ""),
+            ),
+        )
+
+        # returns if to stop, True means stop iterating.
+        def fetch_schema_data(cursor, db_name):
+            db_info = self._query_db_information(db_name, cursor)
+            schemas = self._query_schema_information(cursor)
+            self._dataSubmitter.store_db_info(db_name, db_info)
+            for schema in schemas:
+                tables = self._get_tables(schema, cursor)
+                tables_chunks = list(get_list_chunks(tables, self.TABLES_CHUNK_SIZE))
+                for tables_chunk in tables_chunks:
+                    if self._dataSubmitter.exceeded_total_columns_number():
```

Comment:
Looks like `_dataSubmitter._total_columns_count` will keep increasing during one check run but not decreased when the events in the queue are flushed to backend. I know the integration probably does not have clear insights into the queue status. 
But based on current implementation, if the monitored database has number of tables & columns exceed `MAX_TOTAL_COLUMN_COUNT`, the integration will always submit same set of tables and stop iteration at same position on every check run.

Reply:
that's correct. We will fetch same tables and stop at the same point every time. I have a prototype of a logic that starts from the last table and continues but it wasn't trivial. Basically, I started to code it then quickly realised it will take some time to implement and I wanted to have a first review of an MVP. There is another optimisation which we can add - https://github.com/DataDog/integrations-core/pull/17258/files#r1593532289 . After the MVP is in. Justin proposed opportunistically take next tables in the list (there is no guarantee that table list doesnt change between checks) what do you think ?